In [1]:
# Mengimpor library yang diperlukan
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense

# Inisialisasi CNN
MesinKlasifikasi = Sequential()

# Langkah 1 - Convolution
MesinKlasifikasi.add(Conv2D(filters = 32, kernel_size=(3, 3), input_shape = (128, 128, 3), activation = 'relu'))

# Langkah 2 - Pooling
MesinKlasifikasi.add(MaxPooling2D(pool_size = (2, 2)))

# Menambah convolutional layer
MesinKlasifikasi.add(Conv2D(32, (3, 3), activation = 'relu'))
MesinKlasifikasi.add(MaxPooling2D(pool_size = (2, 2)))

# Langkah 3 - Flattening
MesinKlasifikasi.add(Flatten())

# Langkah 4 - Full connection
MesinKlasifikasi.add(Dense(units = 128, activation = 'relu'))
MesinKlasifikasi.add(Dense(units = 1, activation = 'sigmoid'))

# Menjalankan CNN
MesinKlasifikasi.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])

# Menjalankan CNN ke training dan test set
from tensorflow.keras.preprocessing.image import ImageDataGenerator

train_datagen = ImageDataGenerator(rescale = 1./255,
                                   shear_range = 0.2,
                                   zoom_range = 0.2,
                                   horizontal_flip = True)

test_datagen = ImageDataGenerator(rescale = 1./255)

training_set = train_datagen.flow_from_directory('dataset/training_set',
                                                 target_size = (128, 128),
                                                 batch_size = 32,
                                                 class_mode = 'binary')

test_set = test_datagen.flow_from_directory('dataset/test_set',
                                            target_size = (128, 128),
                                            batch_size = 32,
                                            class_mode = 'binary')

MesinKlasifikasi.fit(training_set,
                         steps_per_epoch = 8000//32,
                         epochs = 50,
                         validation_data = test_set,
                         validation_steps = 2000//32)

c:\Users\ASUS\AppData\Local\Programs\Python\Python311\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Found 8000 images belonging to 2 classes.
Found 2000 images belonging to 2 classes.


c:\Users\ASUS\AppData\Local\Programs\Python\Python311\Lib\site-packages\keras\src\trainers\data_adapters\py_dataset_adapter.py:121: UserWarning: Your `PyDataset` class should call `super().__init__(**kwargs)` in its constructor. `**kwargs` can include `workers`, `use_multiprocessing`, `max_queue_size`. Do not pass these arguments to `fit()`, as they will be ignored.
  self._warn_if_super_not_called()


Epoch 1/50
250/250 ━━━━━━━━━━━━━━━━━━━━ 0s 289ms/step - accuracy: 0.5511 - loss: 0.7716

c:\Users\ASUS\AppData\Local\Programs\Python\Python311\Lib\site-packages\keras\src\trainers\data_adapters\py_dataset_adapter.py:121: UserWarning: Your `PyDataset` class should call `super().__init__(**kwargs)` in its constructor. `**kwargs` can include `workers`, `use_multiprocessing`, `max_queue_size`. Do not pass these arguments to `fit()`, as they will be ignored.
  self._warn_if_super_not_called()


250/250 ━━━━━━━━━━━━━━━━━━━━ 80s 313ms/step - accuracy: 0.5512 - loss: 0.7713 - val_accuracy: 0.5817 - val_loss: 0.6715
Epoch 2/50
250/250 ━━━━━━━━━━━━━━━━━━━━ 0s 496us/step - accuracy: 0.0000e+00 - loss: 0.0000e+00 - val_accuracy: 0.5625 - val_loss: 0.5861
Epoch 3/50


c:\Users\ASUS\AppData\Local\Programs\Python\Python311\Lib\contextlib.py:155: UserWarning: Your input ran out of data; interrupting training. Make sure that your dataset or generator can generate at least `steps_per_epoch * epochs` batches. You may need to use the `.repeat()` function when building your dataset.
  self.gen.throw(typ, value, traceback)


250/250 ━━━━━━━━━━━━━━━━━━━━ 67s 268ms/step - accuracy: 0.6255 - loss: 0.6490 - val_accuracy: 0.6694 - val_loss: 0.6167
Epoch 4/50
250/250 ━━━━━━━━━━━━━━━━━━━━ 0s 254us/step - accuracy: 0.0000e+00 - loss: 0.0000e+00 - val_accuracy: 0.8125 - val_loss: 0.5069
Epoch 5/50
250/250 ━━━━━━━━━━━━━━━━━━━━ 70s 282ms/step - accuracy: 0.6760 - loss: 0.6019 - val_accuracy: 0.6946 - val_loss: 0.5921
Epoch 6/50
250/250 ━━━━━━━━━━━━━━━━━━━━ 0s 256us/step - accuracy: 0.0000e+00 - loss: 0.0000e+00 - val_accuracy: 0.6875 - val_loss: 0.6754
Epoch 7/50
250/250 ━━━━━━━━━━━━━━━━━━━━ 65s 260ms/step - accuracy: 0.7027 - loss: 0.5709 - val_accuracy: 0.7167 - val_loss: 0.5692
Epoch 8/50
250/250 ━━━━━━━━━━━━━━━━━━━━ 0s 196us/step - accuracy: 0.0000e+00 - loss: 0.0000e+00 - val_accuracy: 0.6875 - val_loss: 0.6239
Epoch 9/50
250/250 ━━━━━━━━━━━━━━━━━━━━ 63s 254ms/step - accuracy: 0.7308 - loss: 0.5401 - val_accuracy: 0.6920 - val_loss: 0.5786
Epoch 10/50
250/250 ━━━━━━━━━━━━━━━━━━━━ 0s 216us/step - accuracy: 0.0000

### **Interpretation:**
**1. Impor Library yang Diperlukan**
   Pada awal kode, beberapa pustaka yang diperlukan diimpor:
   - **Sequential**: Digunakan untuk membangun model neural network secara berurutan, di mana setiap lapisan ditambahkan satu per satu.
   - **Conv2D**: Digunakan untuk membuat lapisan konvolusi 2D, yang diperlukan untuk memproses data gambar.
   - **MaxPooling2D**: Digunakan untuk mengurangi dimensi data setelah konvolusi, dengan cara mengambil nilai maksimum dari setiap blok area.
   - **Flatten**: Digunakan untuk mengubah hasil dari lapisan konvolusi menjadi vektor satu dimensi yang bisa diproses lebih lanjut.
   - **Dense**: Digunakan untuk membuat lapisan fully connected dalam model neural network.

**2. Membangun Model CNN**
   Pada baris 6, objek MesinKlasifikasi didefinisikan sebagai model klasifikasi gambar menggunakan Sequential.

   - **Lapisan Convolution (Baris 9)**:
     Menggunakan Conv2D untuk lapisan konvolusi pertama, dengan 32 filter berukuran 3x3, dan menggunakan ReLU sebagai fungsi aktivasi. input_shape diatur ke (128, 128, 3), yang berarti gambar input berukuran 128x128 piksel dan memiliki 3 saluran (RGB).

   - **MaxPooling (Baris 12)**:
     Setelah lapisan konvolusi, digunakan MaxPooling2D dengan ukuran pool 2x2 untuk mengurangi dimensi hasil konvolusi dan mengambil nilai maksimum dalam tiap blok 2x2.

   - **Menambah Lapisan Convolution dan Pooling Lagi (Baris 15-16)**:
     Ditambahkan lagi lapisan konvolusi dengan 32 filter berukuran 3x3 dan diikuti dengan max pooling, untuk menangkap fitur gambar yang lebih dalam dan lebih detail.

   - **Flattening (Baris 19)**:
     Proses Flattening dilakukan untuk mengubah hasil dari lapisan konvolusi dan pooling menjadi vektor satu dimensi yang akan diteruskan ke lapisan fully connected berikutnya.

   - **Fully Connected Layer (Baris 22-23)**:
     Lapisan Dense(128) dengan 128 neuron dan fungsi aktivasi ReLU ditambahkan sebagai lapisan tersembunyi. Pada lapisan output, digunakan Dense(1) dengan fungsi aktivasi sigmoid untuk menghasilkan output biner (0 atau 1), yang cocok untuk masalah klasifikasi biner.

**3. Menyiapkan Data dan Augmentasi**
   - **Image Augmentation**:
     train_datagen dan test_datagen dibuat untuk memproses data pelatihan dan pengujian:
     - train_datagen mengaplikasikan augmentasi gambar seperti rotasi, zoom, dan flip horizontal untuk meningkatkan keragaman data pelatihan dan mengurangi overfitting.
     - test_datagen hanya melakukan rescaling pada data pengujian (konversi piksel menjadi nilai antara 0 dan 1).

   - **Menyiapkan Training dan Test Set**:
     training_set dan test_set didefinisikan menggunakan `flow_from_directory` untuk membaca gambar dari folder yang telah disiapkan. Setiap gambar diubah ukurannya menjadi 128x128 piksel, dengan batch size 32. Data pelatihan diaugmentasi, sedangkan data pengujian hanya di-rescale.

**4. Melatih Model**
   Pada baris 48, model dilatih dengan menggunakan data pelatihan dan divalidasi dengan data pengujian. 
   - steps_per_epoch dihitung dengan membagi jumlah total gambar pelatihan (8000) dengan ukuran batch (32), sehingga diperoleh 250 langkah per epoch.
   - validation_steps dihitung dengan cara yang sama untuk data pengujian (2000 gambar dibagi 32), menghasilkan 62 langkah per epoch.
   - Model dilatih selama 50 epoch, di mana setiap epoch akan mengupdate bobot model berdasarkan data pelatihan dan mengevaluasi akurasi pada data pengujian.

**5. Evaluasi dan Hasil**
   Setelah pelatihan selesai, model menunjukkan nilai loss dan accuracy untuk data pelatihan dan data pengujian (validation). Akurasi dan loss di data pelatihan dan pengujian diharapkan semakin baik seiring bertambahnya epoch, yang menunjukkan bahwa model belajar lebih baik dan dapat menggeneralisasi pola dengan lebih baik.

In [2]:
# Uji coba lagi di test_set
import numpy as np
from keras.preprocessing import image
count_dog = 0
count_cat = 0
for i in range(4001, 5001): 
    test_image = image.load_img('dataset/test_set/dogs/dog.' + str(i) + '.jpg', target_size = (128, 128))
    test_image = image.img_to_array(test_image)
    test_image = np.expand_dims(test_image, axis = 0)
    result = MesinKlasifikasi.predict(test_image)
    training_set.class_indices
    if result[0][0] == 0:
        prediction = 'cat'
        count_cat = count_cat + 1
    else:
        prediction = 'dog'
        count_dog = count_dog + 1

# Mencetak hasil prediksinya agar bisa dibaca
print("count_dog:" + str(count_dog))    
print("count_cat:" + str(count_cat))

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 441ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 47ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 42ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 23ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 18ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 38ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 33ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 35ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 29ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 27ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 44ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 44ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 54ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 34ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 33ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 79ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 63ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 36ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 33ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 40ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 69ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 98ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 50ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 42ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 49ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 46ms/step
1/1 ━━━━━━━

### **Interpretasi:**

Berikut adalah penjelasan kode yang telah disesuaikan:

**1. Impor Library yang Diperlukan**
   - **NumPy**: Digunakan untuk manipulasi array dan data numerik.
   - **Keras image preprocessing**: Modul ini digunakan untuk memuat gambar dan mengonversinya menjadi format yang dapat diterima oleh model.

**2. Inisialisasi Variabel Penghitung**
   - **count_dog dan count_cat**: Dua variabel ini digunakan untuk menghitung jumlah prediksi yang benar untuk kategori 'dog' dan 'cat'.

**3. Proses Prediksi**
   - **Looping (Baris 6-17)**: Loop ini melakukan prediksi terhadap gambar dari 4001 hingga 5000 (ini berarti gambar uji yang digunakan berada di rentang tersebut). 
     - Gambar dimuat menggunakan image.load_img dengan ukuran target (128, 128).
     - image.img_to_array mengonversi gambar yang dimuat menjadi array numpy.
     - np.expand_dims menambah dimensi ekstra untuk memastikan gambar memiliki bentuk yang sesuai dengan input yang diinginkan oleh model CNN (dimensi 4D, termasuk batch size).
     - MesinKlasifikasi.predict(test_image) digunakan untuk mendapatkan hasil prediksi dari model, yang akan menghasilkan nilai berupa array dengan prediksi untuk dua kelas (cat atau dog).
     
   - **Prediksi Kelas (Baris 11)**:
     - Berdasarkan hasil prediksi (`result[0][0]`), jika hasilnya 0, maka gambar diprediksi sebagai 'cat'. Jika hasilnya 1, maka gambar diprediksi sebagai 'dog'.
     - count_cat dan count_dog akan dihitung berdasarkan hasil prediksi yang benar. Jika prediksi adalah 'cat', maka count_cat akan bertambah, dan jika prediksi adalah 'dog', maka count_dog akan bertambah.

**4. Mencetak Hasil Prediksi**
   - Setelah semua gambar diproses, count_dog dan count_cat dicetak untuk menampilkan jumlah gambar yang diprediksi sebagai anjing dan kucing.

**5. Hasil Output**
   - Output dari kode ini adalah jumlah gambar yang diprediksi sebagai 'dog' dan 'cat', yang akan terlihat di console. Sebagai contoh, jika model berhasil memprediksi sebagian besar gambar dengan benar, maka nilai count_dog dan count_cat akan menunjukkan jumlah prediksi yang benar untuk masing-masing kategori.